core/src/main/scala/trees/TreeUtils.scala

```diff
@@ -0,0 +1,161 @@
+        case MacroTypeDecl(tpe, bind) =>
+            rss(tpe).map(MacroTypeDecl(_, bind)) orElse
+            rss(bind).map(MacroTypeDecl(tpe, _))
+        case MacroTypeInst(v1, args) =>
+            def convert(a: (Boolean,Tree)) =  a match{
+                case (a1,a2) => rss(a2).map(newA2 => (a1,newA2))
+            }
+            rss(v1).map(MacroTypeInst(_, args)) orElse
+            mapFirst(args,convert).map(MacroTypeInst(v1,_))
+        case NatMatch(t, t0, bind) =>
+            rss(t).map(NatMatch(_,t0,bind)) orElse
+            rss(t0).map(NatMatch(t,_,bind)) orElse
+            rss(bind).map(NatMatch(t,t0,_))
+        case EitherMatch(t, bind1, bind2) =>
+            rss(t).map(EitherMatch(_,bind1,bind2)) orElse
+            rss(bind1).map(EitherMatch(t,_,bind2)) orElse
+            rss(bind2).map(EitherMatch(t,bind1,_))
+        case Primitive(op, args) =>
+            mapFirst2(args, rss(_)).map(Primitive(op,_))
+        case ErasableApp(t1, t2) =>
+            rss(t1).map(ErasableApp(_,t2)) orElse
+            rss(t2).map(ErasableApp(t1,_))
+        case Fold(tp, t) => 
+            rss(tp).map(Fold(_,t)) orElse
+            rss(t).map(Fold(tp,_))
+        case Unfold(t, bind) =>
+            rss(t).map(Unfold(_,bind)) orElse
+            rss(bind).map(Unfold(t,_))
+        case UnfoldPositive(t, bind) => 
+            rss(t).map(UnfoldPositive(_,bind)) orElse
+            rss(bind).map(UnfoldPositive(t,_))
+        case Abs(t) => 
+            rss(t).map(Abs(_))
+        case TypeApp(t1, t2) => 
+            rss(t1).map(TypeApp(_,t2)) orElse
+            rss(t2).map(TypeApp(t1,_))
+        case Error(s, op) => 
+            op.flatMap(rss(_).map(newT => Error(s,Some(newT))))
+            
+        case NatType => None
+        case BoolType => None
+        case UnitType => None
+
+        case SumType(t1, t2) => 
+            rss(t1).map(SumType(_,t2)) orElse
+            rss(t2).map(SumType(t1,_))
+        case PiType(t1, bind) => 
+            rss(t1).map(PiType(_,bind)) orElse
+            rss(bind).map(PiType(t1,_))
+        case SigmaType(t1, bind) =>
+            rss(t1).map(SigmaType(_,bind)) orElse
+            rss(bind).map(SigmaType(t1,_))
+        case IntersectionType(t1, bind) => 
+            rss(t1).map(IntersectionType(_,bind)) orElse
+            rss(bind).map(IntersectionType(t1,_))
+        case RefinementType(t1, bind) => 
+            rss(t1).map(RefinementType(_,bind)) orElse
+            rss(bind).map(RefinementType(t1,_))
+        case RecType(n, bind) => 
+            rss(n).map(RecType(_,bind)) orElse
+            rss(bind).map(RecType(n,_))
+        case PolyForallType(bind) =>
+            rss(bind).map(PolyForallType(_))
+        
+        case BottomType => None
+        case TopType => None
+
+        case _ => throw new java.lang.Exception(s"Function `replaceSmallStep` is not implemented on $body (${body.getClass}).")
+      }
+    }
+    */
+  }
+  
+  def replaceFirstConditionalOnSuperTree(
```

Comment:
Those two methods could be moved to Trees.scala, although it is pretty full already

Reply:
Yes, functions on trees need some reorganization, we can keep this for later.